# Getting Financial Data - Pandas Datareader

### Introduction:

This time you will get data from a website.


### Step 1. Import the necessary libraries

In [30]:
import pandas as pd
import requests
import numpy as np
from datetime import timedelta
from datetime import datetime
import pandas_datareader.data as web

### Step 2. Create your time range (start and end variables). The start date should be 01/01/2015 and the end should today (whatever your today is).

In [15]:
date_format = "%d/%m/%Y"
inicio = datetime.strptime('01/01/2015', date_format)
fin = datetime.today().strftime('%d-%m-%Y')
fin

'12-02-2022'

### Step 3. Get an API key for one of the APIs that are supported by Pandas Datareader, preferably for AlphaVantage.

If you do not have an API key for any of the supported APIs, it is easiest to get one for [AlphaVantage](https://www.alphavantage.co/support/#api-key). (Note that the API key is shown directly after the signup. You do *not* receive it via e-mail.)

(For a full list of the APIs that are supported by Pandas Datareader, [see here](https://pydata.github.io/pandas-datareader/readers/index.html). As the APIs are provided by third parties, this list may change.)

In [16]:
API = 'BTE9NFB6JDAU4KY1'

### Step 4. Use Pandas Datarader to read the daily time series for the Apple stock (ticker symbol AAPL) between 01/01/2015 and today, assign it to df_apple and print it.

In [33]:
url = 'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=AAPL&apikey=BTE9NFB6JDAU4KY1'
r = requests.get(url)
apple = r.json()
apple = pd.DataFrame.from_dict(apple, orient='index')
apple.head(10)

1. Information  \
Meta Data            Daily Prices (open, high, low, close) and Volumes   
Time Series (Daily)                                                NaN   

                    2. Symbol 3. Last Refreshed 4. Output Size 5. Time Zone  \
Meta Data                AAPL        2022-02-11        Compact   US/Eastern   
Time Series (Daily)       NaN               NaN            NaN          NaN   

                                                            2022-02-11  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '172.3300', '2. high': '173.0800',...   

                                                            2022-02-10  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '174.1400', '2. high': '175.4800',...   

                                                            2022-02-09  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '176.0500', '2. high': '176.6500',...   

                                                            2022-02-08  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '171.7300', '2. high': '175.3500',...   

                                                            2022-02-07  ...  \
Meta Data                                                          NaN  ...   
Time Series (Daily)  {'1. open': '172.8600', '2. high': '173.9458',...  ...   

                                                            2021-10-05  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '139.4900', '2. high': '142.2400',...   

                                                            2021-10-04  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '141.7600', '2. high': '142.2100',...   

                                                            2021-10-01  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '141.9000', '2. high': '142.9200',...   

                                                            2021-09-30  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '143.6600', '2. high': '144.3780',...   

                                                            2021-09-29  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '142.4700', '2. high': '144.4500',...   

                                                            2021-09-28  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '143.2500', '2. high': '144.7500',...   

                                                            2021-09-27  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '145.4700', '2. high': '145.9600',...   

                                                            2021-09-24  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '145.6600', '2. high': '147.4701',...   

                                                            2021-09-23  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '146.6500', '2. high': '147.0800',...   

                                                            2021-09-22  
Meta Data                                                          NaN  
Time Series (Daily)  {'1. open': '144.4500', '2. high': '146.4300',...  

[2 rows x 105 columns]

#### lo que veo es que sólo hay dos filas y 105 columnas y cada una de esas columnas tiene un diccionario dentro. tengo que encontrar la forma de limpiar la info y asignar cada open, high, low, close a una fecha específica. se me ocurre convertir cada uno de los diccionarios que hay para cada fecha en dataframes. me qiedaría con muchismos dataframes

### Step 5. Add a new column "stock" to the dataframe and add the ticker symbol

### Step 6. Repeat the two previous steps for a few other stocks, always creating a new dataframe: Tesla, IBM and Microsoft. (Ticker symbols TSLA, IBM and MSFT.)

In [34]:
url = 'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=IBM&apikey=BTE9NFB6JDAU4KY1'
r = requests.get(url)
ibm = r.json()
ibm = pd.DataFrame.from_dict(ibm, orient='index')
ibm.head(10)

1. Information  \
Meta Data            Daily Prices (open, high, low, close) and Volumes   
Time Series (Daily)                                                NaN   

                    2. Symbol 3. Last Refreshed 4. Output Size 5. Time Zone  \
Meta Data                 IBM        2022-02-11        Compact   US/Eastern   
Time Series (Daily)       NaN               NaN            NaN          NaN   

                                                            2022-02-11  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '133.9000', '2. high': '134.7099',...   

                                                            2022-02-10  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '135.4700', '2. high': '136.5600',...   

                                                            2022-02-09  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '137.8400', '2. high': '138.3500',...   

                                                            2022-02-08  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '137.2300', '2. high': '137.5200',...   

                                                            2022-02-07  ...  \
Meta Data                                                          NaN  ...   
Time Series (Daily)  {'1. open': '137.4500', '2. high': '137.8223',...  ...   

                                                            2021-10-05  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '144.7500', '2. high': '145.0000',...   

                                                            2021-10-04  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '142.7400', '2. high': '146.0000',...   

                                                            2021-10-01  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '141.0000', '2. high': '143.9700',...   

                                                            2021-09-30  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '140.0000', '2. high': '140.5700',...   

                                                            2021-09-29  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '137.7300', '2. high': '139.9300',...   

                                                            2021-09-28  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '139.1700', '2. high': '139.6880',...   

                                                            2021-09-27  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '137.9600', '2. high': '139.0650',...   

                                                            2021-09-24  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '137.0300', '2. high': '138.4800',...   

                                                            2021-09-23  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '135.2500', '2. high': '137.4200',...   

                                                            2021-09-22  
Meta Data                                                          NaN  
Time Series (Daily)  {'1. open': '133.7200', '2. high': '135.3700',...  

[2 rows x 105 columns]

In [35]:
url = 'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=TSLA&apikey=BTE9NFB6JDAU4KY1'
r = requests.get(url)
tesla = r.json()
tesla = pd.DataFrame.from_dict(tesla, orient='index')
tesla.head(10)

1. Information  \
Meta Data            Daily Prices (open, high, low, close) and Volumes   
Time Series (Daily)                                                NaN   

                    2. Symbol 3. Last Refreshed 4. Output Size 5. Time Zone  \
Meta Data                TSLA        2022-02-11        Compact   US/Eastern   
Time Series (Daily)       NaN               NaN            NaN          NaN   

                                                            2022-02-11  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '909.6300', '2. high': '915.9600',...   

                                                            2022-02-10  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '908.3700', '2. high': '943.8100',...   

                                                            2022-02-09  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '935.0000', '2. high': '946.2699',...   

                                                            2022-02-08  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '905.5300', '2. high': '926.2899',...   

                                                            2022-02-07  ...  \
Meta Data                                                          NaN  ...   
Time Series (Daily)  {'1. open': '923.7900', '2. high': '947.7700',...  ...   

                                                            2021-10-05  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '784.7962', '2. high': '797.3100',...   

                                                            2021-10-04  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '796.5000', '2. high': '806.9699',...   

                                                            2021-10-01  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '778.4000', '2. high': '780.7800',...   

                                                            2021-09-30  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '781.0000', '2. high': '789.1305',...   

                                                            2021-09-29  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '779.8000', '2. high': '793.5000',...   

                                                            2021-09-28  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '787.2000', '2. high': '795.6400',...   

                                                            2021-09-27  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '773.1200', '2. high': '799.0000',...   

                                                            2021-09-24  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '745.8900', '2. high': '774.8000',...   

                                                            2021-09-23  \
Meta Data                                                          NaN   
Time Series (Daily)  {'1. open': '755.0000', '2. high': '758.2000',...   

                                                            2021-09-22  
Meta Data                                                          NaN  
Time Series (Daily)  {'1. open': '743.5263', '2. high': '753.6699',...  

[2 rows x 105 columns]

In [38]:
url = 'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY&symbol=MSFT&apikey=BTE9NFB6JDAU4KY1'
r = requests.get(url)
micro = r.json()
micro = pd.DataFrame.from_dict(micro, orient='index')
micro.head(10)

IndentationError: unexpected indent (<ipython-input-38-079e92df9183>, line 5)

### Step 7. Combine the four separate dataFrames into one combined dataFrame df that holds the information for all four stocks

In [37]:
df= apple.append('telsa', 'micro', 'ibm')

TypeError: cannot concatenate object of type '<class 'str'>'; only Series and DataFrame objs are valid

### Step 8. Shift the stock column into the index (making it a multi-level index consisting of the ticker symbol and the date).

### Step 7. Create a dataFrame called vol, with the volume values.

### Step 8. Aggregate the data of volume to weekly.
Hint: Be careful to not sum data from the same week of 2015 and other years.

### Step 9. Find all the volume traded in the year of 2015